worker/src/socket.rs

```diff
@@ -0,0 +1,432 @@
+use std::task::Poll;
+
+use crate::r2::js_object;
+use crate::Result;
+use futures_util::FutureExt;
+use js_sys::{Boolean as JsBoolean, JsString, Number as JsNumber, Object as JsObject, Uint8Array};
+use wasm_bindgen::{JsCast, JsValue};
+use wasm_bindgen_futures::JsFuture;
+
+enum Reading {
+    None,
+    Pending(JsFuture, web_sys::ReadableStreamDefaultReader),
+    Ready(Vec<u8>),
+}
+
+impl Default for Reading {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+enum Writing {
+    Pending(JsFuture, web_sys::WritableStreamDefaultWriter, usize),
+    None,
+}
+
+impl Default for Writing {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+enum Closing {
+    Pending(JsFuture),
+    None,
+}
+
+impl Default for Closing {
+    fn default() -> Self {
+        Self::None
+    }
+}
+
+/// Represents an outbound TCP connection from your Worker.
+pub struct Socket {
+    inner: worker_sys::Socket,
+    writable: web_sys::WritableStream,
+    readable: web_sys::ReadableStream,
+    write: Option<Writing>,
+    read: Option<Reading>,
+    close: Option<Closing>,
+}
+
+// This can only be done because workers are single threaded.
+unsafe impl Send for Socket {}
+unsafe impl Sync for Socket {}
+
+impl Socket {
+    fn new(inner: worker_sys::Socket) -> Self {
+        let writable = inner.writable();
+        let readable = inner.readable();
+        Socket {
+            inner,
+            writable,
+            readable,
+            read: None,
+            write: None,
+            close: None,
+        }
+    }
+
+    /// Closes the TCP socket. Both the readable and writable streams are forcibly closed.
+    pub async fn close(&mut self) -> Result<()> {
+        JsFuture::from(self.inner.close()).await?;
+        Ok(())
+    }
+
+    /// This Future is resolved when the socket is closed
+    /// and is rejected if the socket encounters an error.
+    pub async fn closed(&self) -> Result<()> {
+        JsFuture::from(self.inner.closed()).await?;
+        Ok(())
+    }
+
+    /// Upgrades an insecure socket to a secure one that uses TLS,
+    /// returning a new Socket. Note that in order to call this method,
+    /// you must set [`secure_transport`](SocketOptions::secure_transport)
+    /// to [`StartTls`](SecureTransport::StartTls) when initially
+    /// calling [`connect`](connect) to create the socket.
+    pub fn start_tls(self) -> Socket {
+        let inner = self.inner.start_tls();
+        Socket::new(inner)
+    }
+
+    pub fn builder() -> ConnectionBuilder {
+        ConnectionBuilder::default()
+    }
+
+    fn handle_write_future(
+        cx: &mut std::task::Context<'_>,
```

Comment:
nit: I think this code would be easier to read if we used `use` to import instead of a bunch of qualified paths.